cli/cli-client/src/main/java/org/infinispan/cli/patching/PatchTool.java

```diff
@@ -0,0 +1,474 @@
+      if (installedPatches.isEmpty()) {
+         throw MSG.patchNoPatchesInstalledToRollback();
+      }
+      PatchInfo patchInfo = installedPatches.remove(installedPatches.size() - 1);
+      Path backup = getBackupPath(target, patchInfo);
+      for (PatchOperation operation : patchInfo.getOperations()) {
+         switch (operation.getAction()) {
+            case ADD: {
+               Path file = target.resolve(operation.getNewPath());
+               // Remove any added files
+               println(dryRun, MSG.patchRollbackFile(file));
+               if (!dryRun) {
+                  Files.delete(file);
+               }
+            }
+            break;
+            case SOFT_REPLACE: {
+               // We only restore if the file hasn't been changed
+               Path file = target.resolve(operation.getPath());
+               String sha256 = sha256(file);
+               if (sha256.equals(operation.getDigest())) {
+                  println(dryRun, MSG.patchRollbackFile(file));
+                  if (!dryRun) {
+                     Files.move(backup.resolve(operation.getPath()), target.resolve(operation.getPath()), StandardCopyOption.REPLACE_EXISTING);
+                  }
+               }
+            }
+            case UPGRADE: {
+               Path file = target.resolve(operation.getNewPath());
+               println(dryRun, MSG.patchRollbackFile(file));
+               if (!dryRun) {
+                  Files.delete(file);
+               }
+            }
+            // Fall through to add the backed-up file
+            case REMOVE:
+            case HARD_REPLACE: {
+               Path file = target.resolve(operation.getPath());
+               println(dryRun, MSG.patchRollbackFile(file));
+               if (!dryRun) {
+                  Files.move(backup.resolve(operation.getPath()), target.resolve(operation.getPath()), StandardCopyOption.REPLACE_EXISTING);
+               }
+            }
+            break;
+         }
+      }
+      if (!dryRun) {
+         writeInstalledPatches(target, installedPatches);
+      }
+      println(dryRun, MSG.patchRollback(patchInfo));
+   }
+
+   private void println(boolean dryRun, String msg) {
+      if (dryRun) {
+         out.print(MSG.patchDryRun());
+      }
+      out.println(msg);
+   }
+
+   private Path getBackupPath(Path target, PatchInfo patchInfo) {
+      return target.resolve(PATCHES_DIR).resolve(patchInfo.getSourceVersion() + "_" + patchInfo.getTargetVersion());
+   }
+
+   private List<PatchInfo> getInstalledPatches(Path target) {
+      Path patchesFile = target.resolve(PATCHES_DIR).resolve(PATCHES_FILE);
+      try (InputStream is = Files.newInputStream(patchesFile, StandardOpenOption.READ)) {
+         JavaType patchInfoList = mapper.getTypeFactory().constructCollectionType(List.class, PatchInfo.class);
+         return mapper.readValue(is, patchInfoList);
+      } catch (NoSuchFileException e) {
+         return new ArrayList<>();
```

Comment:
```suggestion
         return Collections.emptyList();
```

Reply:
Nope, because I need to manipulate the list in case of installation.